products/memgraph/upcoming/tutorials/understanding-music-with-modules.md

```diff
@@ -0,0 +1,450 @@
+    g = _create_undirected_graph(context)
+
+    def get_centrality_info(user_id, centrality):
+        return {
+            'user': context.graph.get_vertex_by_id(user_id),
+            'centrality': centrality
+        }
+
+    centralities = nxa.centrality.betweenness_centrality(g)
+    return mgp.Record(
+        centralities=[
+            get_centrality_info(id, ct) for id, ct in centralities.items()
+        ]
+    )
+```
+
+Now let's take a look at the results:
+
+```
+CALL deezer_example.betweenness_centrality()
+YIELD centralities
+UNWIND centralities as c
+RETURN c.user, c.centrality
+ORDER BY c.centrality DESC
+LIMIT 10;
+```
+
+Note: calculating betweenness centrality for each node can take some time to
+finish. The issue of slower `NetworkX` implementations is something we at
+Memgraph would like to address in the future.
+
+For our last trick, let's try to locate communities inside our network.
+Communities are a set of nodes that are densely connected internally. As for
+centrality, there are multiple algorithms for finding communities in a graph. We
+will write a function that takes a method for calculating communities, uses it
+to find the communities, and, optionally, calculates some metrics specific to
+the graph partitioning so we can compare algorithms. To make things more
+interesting, let's find out which genre is the most popular in the community and
+return the percentage which tells us how many of the users have that genre in
+their list. In the end, music is something that connects us!
+
+```python
+import itertools
+
+def _get_communities(context: mgp.ProcCtx,
+                     community_function,
+                     calculate_quality: bool):
+    g = _create_undirected_graph(context)
+
+    communities = list(community_function(g))
+
+    if calculate_quality:
+        quality = {
+            'coverage': nxa.community.quality.coverage(g, communities[:]),
+            'modularity': nxa.community.quality.modularity(g, communities[:]),
+            'performance': nxa.community.quality.performance(g, communities[:])
+        }
+    else:
+        quality = {}
+
+    communities = [
+        list(map(context.graph.get_vertex_by_id, com)) for com in communities
+    ]
+
+    def get_community_info(community):
+        info = {
+            'size': len(community),
+        }
+
+        genre_count = defaultdict(lambda: 0)
+        for genre in itertools.chain(*[u.properties["genres"] for u in community]):
+            genre_count[genre] += 1
+
+        if len(genre_count) != 0:
+            mpg = max(genre_count.items(), key=lambda item: item[1])
+
+            info['most_popular_genre'] = mpg[0]
+            info['most_popular_genre_percentage'] = mpg[1] / info['size']
+
+        return info
+
+    return mgp.Record(
+        communities=[get_community_info(c) for c in communities],
+        quality=quality
+    )
+```
+
+Now that we have our function in place let's test some algorithms. We will be
```

Comment:
Referencing other stuff here is great, but maybe we will need some visual explanation here.